packages/common/src/hooks/useSortedData/useSortedData.ts

```diff
@@ -0,0 +1,56 @@
+import { SortRule, SortBy } from '@openmsupply-client/common';
+import { ObjectWithStringKeys } from './../../types/utility';
+import { useEffect, useState } from 'react';
+import { useSortBy } from '../useSortBy';
+
+const parseValue = (object: any, key: string) => {
+  const value = object[key];
+  if (typeof value === 'string') {
+    const valueAsNumber = Number.parseFloat(value);
+
+    if (!Number.isNaN(valueAsNumber)) return valueAsNumber;
+    return value.toUpperCase(); // ignore case
+  }
+  return value;
+};
+
+const getDataSorter = (sortKey: string, desc: boolean) => (a: any, b: any) => {
+  const valueA = parseValue(a, sortKey);
+  const valueB = parseValue(b, sortKey);
+
+  if (valueA < valueB) {
+    return desc ? 1 : -1;
+  }
+  if (valueA > valueB) {
+    return desc ? -1 : 1;
+  }
+
+  return 0;
+};
+
+interface SortedDataState<T extends ObjectWithStringKeys> {
+  sortedData: T[];
+  sortBy: SortBy<T>;
+  onChangeSortBy: (newSortRule: SortRule<T>) => void;
+}
+
+export const useSortedData = <T extends Record<string, unknown>>(
+  data: T[],
+  initialSortBy: SortRule<T>
+): SortedDataState<T> => {
+  const { sortBy, onChangeSortBy } = useSortBy(initialSortBy);
+  const [sortedData, setSortedData] = useState(data);
+
+  const wrapped = (newSortRule: SortRule<T>) => {
+    const newSortBy = onChangeSortBy(newSortRule);
+    const sorter = getDataSorter(newSortBy.key, !!newSortBy.isDesc);
+    setSortedData(data.sort(sorter));
+  };
+
+  useEffect(() => {
+    const sorter = getDataSorter(sortBy.key, !!sortBy.isDesc);
+    setSortedData(data.sort(sorter));
+  }, []);
+
+  return { sortedData, sortBy, onChangeSortBy: wrapped };
```

Comment:
This is a bit lame.

I started with just using `onChangeSortBy`, which changes `sortBy`, and then putting the `sortBy` in the `useEffect`.. so whenever the `sortBy` changes, resort the data. This was easy and it was nice. But it causes two renders.. one to set the `sortBy` and another to change the `data`, so I went with wrapping the `onChangeSortBy` and doing both in the same callback (which react then batches)

Reply:
Also need to determine the actual data type from the column, to sort by. At the moment we're just inferring the sort, but that's not going to work for long.

I had a play around with adding a `defaultSortDirection` to the column definition, and then when a trigger is sorted, pass the whole column to the call back.

This worked well but was a bit too much to add into this PR

Reply:
good spotting of the double render. I had thought to add the `defaultSortDirection` and something for the datatype to the column definition. yeah, next PR for that.

Reply:
Though the double render is probably irrelevant 🤔